1-Draft/RFC00XX-Module-Isolation.md

```diff
@@ -0,0 +1,251 @@
+if the module that depends on the lower version is loaded first,
+the other module cannot be loaded in the same session anymore.
+
+The proposal is to allow Loading a module into a custom `AssemblyLoadContext`.
+All the module required assemblies, all the assemblies loaded by `Add-Type` in the module script,
+as well as assemblies produced by the PowerShell classes declared in the module,
+will be loaded into the custom load context.
+
+## Motivation
+
+    As a PowerShell user, I'm able to use modules that depend on different versions of the same assembly in the same session.
+
+## Specification
+
+1. Create the type `PSModuleLoadContext` that derives from `AssemblyLoadContext` (ALC).
+   This type will be the custom ALC used for the isolated modules.
+
+1. Update `Import-Module` with two new parameters:
+
+    ```powershell
+    ## Import the module to an isolated AssemblyLoadContext.
+    Import-Module [-Isolated]
+
+    ## Import the module to the specified AssemblyLoadContext.
+    Import-Module [[-AssemblyLoadContext] <PSModuleLoadContext>]
+    ```
+    - When `-Isolated` is specified,
+      PowerShell creates a custom `AssemblyLoadContext` instance of the type `PSModuleLoadContext`,
+      which will be associated with the module.
+
+    - When `-AssemblyLoadContext` is specified,
+      the given `PSModuleLoadContext` instance will be associated with the module.
+
+    Unless otherwise instructed by the module author,
+    all assembly loading triggered by loading the module or executing the module
+    will have the assembly loaded into the associated assembly load context.
+
+    > NOTE: A module author always can explicitly choose where to load an assembly, by:
+    > - calling `LoadXXX` methods directly on an `AssemblyLoadContext` instance;
+    > - changing the active assembly load context with `AssemblyLoadContext.EnterContextualReflection`;
+    > - calling `Assembly.LoadFrom` to load an assembly to the default context;
+    > - calling `Assembly.LoadFile` to load an assembly to a separate anonymous load context;
+
+1. The `ModuleInfo` object will have a new public read-only property named `ReflectionContext`.
+   It will hold a reference to the `PSModuleLoadContext` instance associated with the module.
+
+    > NOTE: The `SessionState` or `SessionStateInternal` was considered when thinking about where to store the custom `PSModuleLoadContext` instance.
+    > It was not chosen because a binary module does not have a `SessionStateInternal` created for it.
+
+1. Nested modules should use the same ALC as the parent module.
+   Nested modules belong to the parent module and commands are also exposed by the parent modules,
+   so they should share the same ALC as the parent module.
+
+1. Required modules are loaded into the default ALC as is today.
+   The comment to the `RequiredModules` key in module manifest is:
+   > Modules that must be imported into the global environment prior to importing this module.
```

Comment:
I'd argue that importing a module into the global environment isn't the same as importing its assembly dependencies...

Reply:
When we use an ALC, we still import modules into the global context (they are visible and usable in the global environment). But the underlying assembly or any dependencies might not be globally available.

Reply:
This is a good point and I have responded to the same concern in https://github.com/PowerShell/PowerShell-RFC/pull/298#discussion_r689869081.